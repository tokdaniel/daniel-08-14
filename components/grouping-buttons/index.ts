export { default as GroupingButtons } from './grouping-buttons'
