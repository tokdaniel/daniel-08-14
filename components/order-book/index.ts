export { default as OrderBook } from './order-book'
export { default as HalfBook } from './half-book'
