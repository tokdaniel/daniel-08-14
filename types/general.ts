export type Delta = [number, number][]
export type IncomingMessage = {
  event?: string
  feed: string
  product_ids: string[]
  asks?: Delta
  bids?: Delta
}

export type Book = {
  asks: Delta
  bids: Delta
}

export type Page = {
  [x: number]: number
}

export type NBook = {
  asks: Page
  bids: Page
}

export type Side = 'bids' | 'asks'
