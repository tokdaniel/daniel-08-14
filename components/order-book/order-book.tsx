import { useCallback, useState } from 'react'
import styled from 'styled-components'

import { Loader } from 'components/loader'
import { Spread } from 'components/spread'
import { Text } from 'components/typography'
import useOrderBook from 'util/hooks/use-order-book'

import HalfBook from './half-book'

const Container = styled.div`
  display: flex;
  max-width: 600px;
  flex-direction: column;
  align-items: center;
  background: ${({ theme }) => theme.palette.bg[400]};
  padding: 20px;
  border-radius: 12px;
`

const Wrapper = styled.div`
  display: flex;
  width: 100%;
  justify-content: center;
  flex-direction: column;
  text-align: center;
  border-bottom: 1px solid ${({ theme }) => theme.palette.grey[100]};
`

const groupings = [1, 2, 5, 10, 20, 50, 100, 200, 500, 1000, 2000]

const OrderBook: React.FC = () => {
  const [grouping, setGrouping] = useState(groupings[6])
  const book = useOrderBook(grouping)

  const incrementGrouping = useCallback(() => {
    const i = groupings.indexOf(grouping)
    if (i === groupings.length - 1) {
      return
    }
    setGrouping(groupings[i + 1])
  }, [grouping])

  const decrementGrouping = useCallback(() => {
    const i = groupings.indexOf(grouping)
    if (i === 0) {
      return
    }
    setGrouping(groupings[i - 1])
  }, [grouping])

  return book ? (
    <Container>
      <Text variant="h2">XBT/USD</Text>
      <Wrapper>
        <HalfBook items={book.bids} side="bids" />
        <HalfBook items={book.asks} side="asks" />
      </Wrapper>
      <Spread
        book={book}
        grouping={grouping}
        incrementGrouping={incrementGrouping}
        decrementGrouping={decrementGrouping}
      />
    </Container>
  ) : (
    <Loader
      alt="Loading..."
      width={150}
      height={150}
      className="loader-image"
    />
  )
}

export default OrderBook
