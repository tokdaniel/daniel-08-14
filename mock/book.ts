import { Book } from 'types/general'

export const simpleBook: Book = {
  bids: [
    [47599.5, 171],
    [47597, 1070],
    [47592, 4479],
  ],
  asks: [
    [47602.5, 2500],
    [47603, 1808],
    [47610, 2500],
  ],
}

export const book: Book = {
  bids: [
    [47599.5, 171],
    [47597, 1070],
    [47592, 4479],
    [47590.5, 170496],
    [47590, 5819],
    [47589.5, 2031],
    [47588, 2500],
    [47586, 26700],
    [47583, 11553],
    [47582.5, 10000],
    [47581.5, 4388],
    [47581, 3862],
    [47580, 57718],
    [47579.5, 3351],
    [47572.5, 6138],
    [47569.5, 4324],
    [47569, 62438],
    [47568.5, 203737],
    [47567, 10000],
    [47566, 2093],
    [47563, 10000],
    [47562.5, 20000],
    [47561.5, 10396],
    [47561, 11],
    [47559, 20000],
  ],
  asks: [
    [47602.5, 2500],
    [47603, 1808],
    [47610, 2500],
    [47610.5, 4043],
    [47613, 2500],
    [47614, 1986],
    [47614.5, 2500],
    [47615.5, 1070],
    [47617, 4637],
    [47617.5, 5315],
    [47622, 4565],
    [47622.5, 10000],
    [47623, 2500],
    [47623.5, 16138],
    [47626, 63948],
    [47626.5, 26700],
    [47627.5, 4737],
    [47628.5, 5000],
    [47629.5, 1976],
    [47637.5, 199658],
    [47638, 3026],
    [47638.5, 2982],
    [47641.5, 38081],
    [47642.5, 1937],
    [47644, 11365],
  ],
}
