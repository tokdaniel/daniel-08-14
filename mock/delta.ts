import { Book } from 'types/general'

export const delta: Book = {
  bids: [
    [47599.5, 0],
    [47597, 1259],
    [47592, 3345],
  ],
  asks: [
    [47602.5, 0],
    [47603, 2239],
    [47610, 1925],
  ],
}
