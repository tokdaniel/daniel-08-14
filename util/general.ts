import { Book, NBook, Page, Side } from 'types/general'

export const transformSide = (state: Book, side: 'bids' | 'asks'): Page => {
  return Object.values(state[side] ?? {}).reduce((acc, [price, size]) => {
    return {
      ...acc,
      [price]: size,
    }
  }, {})
}

export const transformBook = (book: Book): NBook => {
  const asks = transformSide(book, 'asks')
  const bids = transformSide(book, 'bids')

  return {
    asks,
    bids,
  }
}

export const sortBySide =
  (side: Side) =>
  ([priceA]: [string, number], [priceB]: [string, number]): number => {
    const a = parseFloat(priceA)
    const b = parseFloat(priceB)

    return side === 'asks' ? a - b : b - a
  }

export const computeDelta = (state: Page, delta: Page): Page =>
  Object.entries(Object.assign({}, state, delta))
    .filter(([_, size]) => size !== 0)
    .reduce(
      (acc, [price, size]) => ({
        ...acc,
        [price]: size,
      }),
      {},
    )

type Spread = {
  percentage: string
  value: number
}

export const calculateSpread = (minAsk: number, maxBid: number): Spread => {
  const value = minAsk - maxBid
  const percentage = (1 - minAsk / maxBid) * 100

  return { value, percentage: Math.abs(percentage).toFixed(3) }
}

export const getMinAsk = (asks: Page): number =>
  Math.min(...(Object.keys(asks) as unknown as number[]))

export const getMaxBid = (bids: Page): number =>
  Math.max(...(Object.keys(bids) as unknown as number[]))
