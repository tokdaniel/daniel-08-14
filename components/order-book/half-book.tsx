import React, { useMemo, useRef } from 'react'
import styled from 'styled-components'

import { Text } from 'components/typography'
import { NBook } from 'types/general'
import { sortBySide } from 'util/general'

const Container = styled.div`
  display: flex;
  width: 100%;
  flex-direction: column;
  padding: 8px;
`

const OrderBookRow = styled.div<{ header: boolean }>`
  display: flex;
  flex-direction: row;
  height: 20;
  justify-content: space-around;
  align-items: center;
  padding: 2px 0;
  margin-bottom: ${({ header }) => (header ? '8px' : 'initial')};

  ${Text} {
    text-align: right;
  }
`

const Price = styled.div<{ side?: 'asks' | 'bids' }>`
  text-transform: uppercase;
  flex: 1;
  color: ${({ side, theme }) => {
    if (!side) {
      return theme.palette.grey[900]
    }

    return side === 'asks' ? theme.palette.red[500] : theme.palette.green[500]
  }};
`

const Size = styled.div`
  text-transform: uppercase;
  flex: 1;
`

const Total = styled.div`
  text-transform: uppercase;
  flex: 1;
`

type Props = {
  items: NBook['bids']
  side: 'bids' | 'asks'
}

const HalfBook: React.FC<Props> = ({ items, side }) => {
  const total = useRef(0)
  const content = useMemo(
    () =>
      Object.entries(items)
        .sort(sortBySide(side))
        .slice(0, 8)
        .map(([price, size], i) => {
          total.current = i === 0 ? size : total.current + size
          return (
            <OrderBookRow header={false} key={price}>
              <Price side={side}>
                <Text
                  variant="body-1"
                  textColor={
                    side === 'asks' ? ['red', '500'] : ['green', '500']
                  }
                >
                  {price}
                </Text>
              </Price>
              <Size>
                <Text variant="body-1">{size}</Text>
              </Size>
              <Total>
                <Text variant="body-1">{total.current}</Text>
              </Total>
            </OrderBookRow>
          )
        }),
    [items],
  )

  return (
    <Container>
      <OrderBookRow header>
        <Price>
          <Text variant="subtitle-1">price</Text>
        </Price>
        <Size>
          <Text variant="subtitle-1">size</Text>
        </Size>
        <Total>
          <Text variant="subtitle-1">total</Text>
        </Total>
      </OrderBookRow>
      {content}
    </Container>
  )
}

export default HalfBook
