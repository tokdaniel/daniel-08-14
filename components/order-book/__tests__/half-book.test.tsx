import React from 'react'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'

import { HalfBook } from 'components/order-book'
import theme from 'config/theme.config'
import { book } from 'mock/book'
import { transformBook } from 'util/general'

describe('HalfBook', () => {
  it('renders asks correctly', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={theme}>
          <HalfBook items={transformBook(book).asks} side="asks" />,
        </ThemeProvider>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders bids correctly', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={theme}>
          <HalfBook items={transformBook(book).bids} side="bids" />,
        </ThemeProvider>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
