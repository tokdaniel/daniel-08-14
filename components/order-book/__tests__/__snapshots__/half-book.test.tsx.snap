// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HalfBook renders asks correctly 1`] = `
Array [
  <div
    className="sc-gKAaRy cPTABi"
  >
    <div
      className="sc-iCoGMd dZYJlD"
    >
      <div
        className="sc-fujyAs dbumup"
      >
        <div
          className="sc-bdnxRM fmZhhZ"
        >
          price
        </div>
      </div>
      <div
        className="sc-pNWdM iJTsUG"
      >
        <div
          className="sc-bdnxRM fmZhhZ"
        >
          size
        </div>
      </div>
      <div
        className="sc-jrsJWt gpLwsY"
      >
        <div
          className="sc-bdnxRM fmZhhZ"
        >
          total
        </div>
      </div>
    </div>
    <div
      className="sc-iCoGMd crrWzt"
    >
      <div
        className="sc-fujyAs lowjsu"
      >
        <div
          className="sc-bdnxRM cBeHvK"
        >
          47602.5
        </div>
      </div>
      <div
        className="sc-pNWdM iJTsUG"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          2500
        </div>
      </div>
      <div
        className="sc-jrsJWt gpLwsY"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          2500
        </div>
      </div>
    </div>
    <div
      className="sc-iCoGMd crrWzt"
    >
      <div
        className="sc-fujyAs lowjsu"
      >
        <div
          className="sc-bdnxRM cBeHvK"
        >
          47603
        </div>
      </div>
      <div
        className="sc-pNWdM iJTsUG"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          1808
        </div>
      </div>
      <div
        className="sc-jrsJWt gpLwsY"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          4308
        </div>
      </div>
    </div>
    <div
      className="sc-iCoGMd crrWzt"
    >
      <div
        className="sc-fujyAs lowjsu"
      >
        <div
          className="sc-bdnxRM cBeHvK"
        >
          47610
        </div>
      </div>
      <div
        className="sc-pNWdM iJTsUG"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          2500
        </div>
      </div>
      <div
        className="sc-jrsJWt gpLwsY"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          6808
        </div>
      </div>
    </div>
    <div
      className="sc-iCoGMd crrWzt"
    >
      <div
        className="sc-fujyAs lowjsu"
      >
        <div
          className="sc-bdnxRM cBeHvK"
        >
          47610.5
        </div>
      </div>
      <div
        className="sc-pNWdM iJTsUG"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          4043
        </div>
      </div>
      <div
        className="sc-jrsJWt gpLwsY"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          10851
        </div>
      </div>
    </div>
    <div
      className="sc-iCoGMd crrWzt"
    >
      <div
        className="sc-fujyAs lowjsu"
      >
        <div
          className="sc-bdnxRM cBeHvK"
        >
          47613
        </div>
      </div>
      <div
        className="sc-pNWdM iJTsUG"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          2500
        </div>
      </div>
      <div
        className="sc-jrsJWt gpLwsY"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          13351
        </div>
      </div>
    </div>
    <div
      className="sc-iCoGMd crrWzt"
    >
      <div
        className="sc-fujyAs lowjsu"
      >
        <div
          className="sc-bdnxRM cBeHvK"
        >
          47614
        </div>
      </div>
      <div
        className="sc-pNWdM iJTsUG"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          1986
        </div>
      </div>
      <div
        className="sc-jrsJWt gpLwsY"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          15337
        </div>
      </div>
    </div>
    <div
      className="sc-iCoGMd crrWzt"
    >
      <div
        className="sc-fujyAs lowjsu"
      >
        <div
          className="sc-bdnxRM cBeHvK"
        >
          47614.5
        </div>
      </div>
      <div
        className="sc-pNWdM iJTsUG"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          2500
        </div>
      </div>
      <div
        className="sc-jrsJWt gpLwsY"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          17837
        </div>
      </div>
    </div>
    <div
      className="sc-iCoGMd crrWzt"
    >
      <div
        className="sc-fujyAs lowjsu"
      >
        <div
          className="sc-bdnxRM cBeHvK"
        >
          47615.5
        </div>
      </div>
      <div
        className="sc-pNWdM iJTsUG"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          1070
        </div>
      </div>
      <div
        className="sc-jrsJWt gpLwsY"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          18907
        </div>
      </div>
    </div>
  </div>,
  ",",
]
`;

exports[`HalfBook renders bids correctly 1`] = `
Array [
  <div
    className="sc-gKAaRy cPTABi"
  >
    <div
      className="sc-iCoGMd dZYJlD"
    >
      <div
        className="sc-fujyAs dbumup"
      >
        <div
          className="sc-bdnxRM fmZhhZ"
        >
          price
        </div>
      </div>
      <div
        className="sc-pNWdM iJTsUG"
      >
        <div
          className="sc-bdnxRM fmZhhZ"
        >
          size
        </div>
      </div>
      <div
        className="sc-jrsJWt gpLwsY"
      >
        <div
          className="sc-bdnxRM fmZhhZ"
        >
          total
        </div>
      </div>
    </div>
    <div
      className="sc-iCoGMd crrWzt"
    >
      <div
        className="sc-fujyAs cGFeBn"
      >
        <div
          className="sc-bdnxRM jjbfsL"
        >
          47599.5
        </div>
      </div>
      <div
        className="sc-pNWdM iJTsUG"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          171
        </div>
      </div>
      <div
        className="sc-jrsJWt gpLwsY"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          171
        </div>
      </div>
    </div>
    <div
      className="sc-iCoGMd crrWzt"
    >
      <div
        className="sc-fujyAs cGFeBn"
      >
        <div
          className="sc-bdnxRM jjbfsL"
        >
          47597
        </div>
      </div>
      <div
        className="sc-pNWdM iJTsUG"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          1070
        </div>
      </div>
      <div
        className="sc-jrsJWt gpLwsY"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          1241
        </div>
      </div>
    </div>
    <div
      className="sc-iCoGMd crrWzt"
    >
      <div
        className="sc-fujyAs cGFeBn"
      >
        <div
          className="sc-bdnxRM jjbfsL"
        >
          47592
        </div>
      </div>
      <div
        className="sc-pNWdM iJTsUG"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          4479
        </div>
      </div>
      <div
        className="sc-jrsJWt gpLwsY"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          5720
        </div>
      </div>
    </div>
    <div
      className="sc-iCoGMd crrWzt"
    >
      <div
        className="sc-fujyAs cGFeBn"
      >
        <div
          className="sc-bdnxRM jjbfsL"
        >
          47590.5
        </div>
      </div>
      <div
        className="sc-pNWdM iJTsUG"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          170496
        </div>
      </div>
      <div
        className="sc-jrsJWt gpLwsY"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          176216
        </div>
      </div>
    </div>
    <div
      className="sc-iCoGMd crrWzt"
    >
      <div
        className="sc-fujyAs cGFeBn"
      >
        <div
          className="sc-bdnxRM jjbfsL"
        >
          47590
        </div>
      </div>
      <div
        className="sc-pNWdM iJTsUG"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          5819
        </div>
      </div>
      <div
        className="sc-jrsJWt gpLwsY"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          182035
        </div>
      </div>
    </div>
    <div
      className="sc-iCoGMd crrWzt"
    >
      <div
        className="sc-fujyAs cGFeBn"
      >
        <div
          className="sc-bdnxRM jjbfsL"
        >
          47589.5
        </div>
      </div>
      <div
        className="sc-pNWdM iJTsUG"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          2031
        </div>
      </div>
      <div
        className="sc-jrsJWt gpLwsY"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          184066
        </div>
      </div>
    </div>
    <div
      className="sc-iCoGMd crrWzt"
    >
      <div
        className="sc-fujyAs cGFeBn"
      >
        <div
          className="sc-bdnxRM jjbfsL"
        >
          47588
        </div>
      </div>
      <div
        className="sc-pNWdM iJTsUG"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          2500
        </div>
      </div>
      <div
        className="sc-jrsJWt gpLwsY"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          186566
        </div>
      </div>
    </div>
    <div
      className="sc-iCoGMd crrWzt"
    >
      <div
        className="sc-fujyAs cGFeBn"
      >
        <div
          className="sc-bdnxRM jjbfsL"
        >
          47586
        </div>
      </div>
      <div
        className="sc-pNWdM iJTsUG"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          26700
        </div>
      </div>
      <div
        className="sc-jrsJWt gpLwsY"
      >
        <div
          className="sc-bdnxRM cLcjEa"
        >
          213266
        </div>
      </div>
    </div>
  </div>,
  ",",
]
`;
